File: artifacts/optipng-0.7.8/decompose/src/z_stream.c
typedef struct z_stream_s
{
  Bytef *next_in;
  uInt avail_in;
  uLong total_in;
  Bytef *next_out;
  uInt avail_out;
  uLong total_out;
  char *msg;
  struct internal_state *state;
  alloc_func zalloc;
  free_func zfree;
  voidpf opaque;
  int data_type;
  uLong adler;
  uLong reserved;
} z_stream
/*** DEPENDENCIES:
typedef unsigned long uLong
----------------------------
typedef Byte Bytef
----------------------------
typedef void *voidp
----------------------------
typedef voidpf (*alloc_func)(voidpf opaque, uInt items, uInt size)
----------------------------
typedef void (*free_func)(voidpf opaque, voidpf address)
----------------------------
typedef unsigned int uInt
----------------------------
typedef void *voidpf
----------------------------
typedef unsigned char Byte
----------------------------
***/

File: artifacts/lodepng/manual_translation/src/z_stream.rs
use crate::*;
use std::os::raw::c_void;


pub struct InternalState; // Opaque type for internal state

pub struct ZStream {
    pub next_in: Option<Box<Bytef>>,
    pub avail_in: uInt,
    pub total_in: uLong,
    pub next_out: Option<Box<Bytef>>,
    pub avail_out: uInt,
    pub total_out: uLong,
    pub msg: Option<Box<i8>>,
    pub state: Option<Box<InternalState>>,
    pub zalloc: AllocFunc,
    pub zfree: FreeFunc,
    pub opaque: voidpf,
    pub data_type: i32,
    pub adler: uLong,
    pub reserved: uLong,
}

pub type Bytef = u8;
pub type uInt = u32;
pub type uLong = u64;
pub type AllocFunc = fn(opaque: voidpf, items: uInt, size: uInt) -> voidpf;
pub type voidpf = *mut c_void;
pub type FreeFunc = Option<Box<dyn Fn(voidpf, voidpf)>>;

